### PR TITLE
[2371] Improve color management in View

### LINKED ISSUE #2371: Improve color management in View
* [x] **I have checked that this feature has not yet been suggested by someone else.**

### Expected behavior

- Default color palettes are provided along with the creation of a view.
-- theme colors
-- mui colors
-- special colors (transparent, none, inherit)
- Handling of new colors on frontend side
- Add new validation rules about colors for View diagrams & forms

### Cuttings backs

...

### Potential side effects

...

### Describe alternatives you've considered

...

### PR DESCRIPTION
**Pull request by AxelRICHARD**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2371

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

**Comment by frouene** [comment]
I think it's missing the impact for the group in form in forms/frontend/sirius-components-forms/src/groups/Group.tsx for the `borderColor`.
And for the widget ref  forms/frontend/sirius-components-widget-reference/src/components/ValuedReferenceAutocomplete.tsx and forms/frontend/sirius-components-widget-reference/src/ReferencePreview.tsx for `color`.

**Comment by frouene** [comment]
What is the expected behavior for the color `none` ?
Currently, when you changed for none it keeps the previous color but after a refresh it's black (test made for a ButtonDescriptionStyle#ForegroundColor in the FormDescriptionEditor).

**Comment by AxelRICHARD** [comment]
> I think it's missing the impact for the group in form in forms/frontend/sirius-components-forms/src/groups/Group.tsx for the `borderColor`. And for the widget ref forms/frontend/sirius-components-widget-reference/src/components/ValuedReferenceAutocomplete.tsx and forms/frontend/sirius-components-widget-reference/src/ReferencePreview.tsx for `color`.

You're right, thank you!

**Comment by AxelRICHARD** [comment]
> What is the expected behavior for the color `none` ? Currently, when you changed for none it keeps the previous color but after a refresh it's black (test made for a ButtonDescriptionStyle#ForegroundColor in the FormDescriptionEditor).

It shouldn't be here. I will delete this value.

File: packages/core/frontend/sirius-components-core/src/color/getCSSColor.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  orange: orange,
+  pink: pink,
+  purple: purple,
+  red: red,
+  teal: teal,
+  yellow: yellow,
+};
+
+export const getCSSColor = (value: string, theme: Theme): string | null => {
+  let cssColor: string | null = null;
```
Comment (sbegaudeau) [inline packages/core/frontend/sirius-components-core/src/color/getCSSColor.ts]: You could start by initializing the color to value and ensure that you would never return `null`, it removes tons of issues

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/Label.tsx [hunk 2]
```diff
@@ -11,26 +11,34 @@
...
 
-const labelStyle = (style: React.CSSProperties, faded: Boolean, transform: string): React.CSSProperties => {
+const labelStyle = (
+  theme: Theme,
+  style: React.CSSProperties,
+  faded: Boolean,
+  transform: string
+): React.CSSProperties => {
   return {
+    ...style,
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/Label.tsx]: This should not be done, it overrides tons of options specified by the backend

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx [hunk 3]
```diff
@@ -25,7 +25,12 @@ const imageNodeStyle = (
   selected: boolean,
   faded: boolean
 ): React.CSSProperties => {
-  const imageNodeStyle: React.CSSProperties = { width: '100%', height: '100%', opacity: faded ? '0.4' : '', ...style };
+  const imageNodeStyle: React.CSSProperties = {
+    ...style,
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ImageNode.tsx]: Style has to be last on purpose

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/edge/MultiLabelEdge.tsx [hunk 4]
```diff
@@ -26,8 +27,9 @@ const multiLabelEdgeStyle = (
   faded: boolean | undefined
 ): React.CSSProperties => {
   const multiLabelEdgeStyle: React.CSSProperties = {
-    opacity: faded ? '0.4' : '',
     ...style,
+    opacity: faded ? '0.4' : '',
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/edge/MultiLabelEdge.tsx:40]: style has to be after opacity on purpose

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ListNode.tsx [hunk 5]
```diff
@@ -26,13 +27,15 @@ const listNodeStyle = (
   faded: boolean
 ): React.CSSProperties => {
   const listNodeStyle: React.CSSProperties = {
+    ...style,
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ListNode.tsx:30]: style has to be after all the default on purpose

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/RectangularNode.tsx [hunk 6]
```diff
@@ -26,12 +27,13 @@ const rectangularNodeStyle = (
   faded: boolean
 ): React.CSSProperties => {
   const rectangularNodeStyle: React.CSSProperties = {
+    ...style,
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/RectangularNode.tsx]: style has to be after the default values

File: packages/sirius-web/backend/sirius-web-services/src/main/java/org/eclipse/sirius/web/services/editingcontext/DynamicRepresentationDescriptionService.java [hunk 7]
```diff
@@ -71,6 +76,8 @@ public List<IRepresentationDescription> findDynamicRepresentationDescriptions(St
             List<View> views = new ArrayList<>();
             List<EPackage> accessibleEPackages = this.getAccessibleEPackages(editingDomain);
             ResourceSet resourceSet = this.createResourceSet(this.ePackageRegistry);
+            this.loadStudioColorPalettes(resourceSet);
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-services/src/main/java/org/eclipse/sirius/web/services/editingcontext/DynamicRepresentationDescriptionService.java:79]: @AxelRICHARD this was the missing piece of code for your information :)

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**